### PR TITLE
gh-115392: Fix doctest reporting incorrect line numbers for decorated functions

### PR DESCRIPTION
**Pull request by brianschubert**
Fixes line numbers reported by `doctest` for wrapper functions (such as those returned by `functools.wraps`).

Currently, `doctest.DocTestFinder._find_lineno` looks at `__code__.co_firstlineno`  to determine the line number that a function's docstring starts at.  This is incorrect for wrapper functions, where the location of the function definition is not the same as the location of the docstring being tested.

Wrapper functions that have been updated with `functools.update_wrapper` have a `__wrapped__` attribute that points to the underlying wrapped function.  This PR adds a recursive check for `__wrapped__` attributes on function objects to locate the original function that defined the docstring-under-test.

* Issue: gh-115392

### CODE REVIEW

File: Lib/doctest.py [hunk 1]
```diff
@@ -1140,6 +1140,8 @@ def _find_lineno(self, obj, source_lines):
             obj = obj.fget
         if inspect.isfunction(obj) and getattr(obj, '__doc__', None):
             # We don't use `docstring` var here, because `obj` can be changed.
+            while hasattr(obj, "__wrapped__"):
+                obj = obj.__wrapped__
```
Comment (AlexWaygood) [inline Lib/doctest.py]: `doctest` already has a heavy dependency on the `inspect` module -- consider using [`inspect.unwrap()`](https://docs.python.org/3/library/inspect.html#inspect.unwrap) here

File: Lib/test/test_doctest/_support.py
Comment (sobolevn) [inline Lib/test/test_doctest/_support.py:1]: `support` usually means something different in our tests, consider naming this file `decorator_mod`

File: Lib/doctest.py (hunk 1 above)
Comment (brianschubert) [reply Lib/doctest.py]: `inspect.unwrap` fits perfectly, thanks! Made the change.

File: Lib/test/test_doctest/_support.py
Comment (brianschubert) [reply Lib/test/test_doctest/_support.py:1]: Good to know. The suggested name sounds good, made the change.

**Review by AlexWaygood — APPROVED** [review]
LGTM. I'm not sure what would happen if you had a class decorator that used `functools.wraps()`, but I can't think of a realistic reason why you might do that. So I don't think that's worth worrying about right now.

**Review by sobolevn — APPROVED** [review]
Thank you!

**Comment by lpsinger** [comment]
This broke doctest discovery for functions that are wrapped but do not define a __doc__. See https://github.com/scientific-python/pytest-doctestplus/pull/248.

**Comment by AlexWaygood** [comment]
> This broke doctest discovery for functions that are wrapped but do not define a `__doc__`. See [scientific-python/pytest-doctestplus#248](https://github.com/scientific-python/pytest-doctestplus/pull/248).

Would you be able to open a new issue with a minimal repro?

**Comment by lpsinger** [comment]
> > This broke doctest discovery for functions that are wrapped but do not define a `__doc__`. See [scientific-python/pytest-doctestplus#248](https://github.com/scientific-python/pytest-doctestplus/pull/248).
> 
> Would you be able to open a new issue with a minimal repro?

https://github.com/python/cpython/issues/117692

**Pull request merged**